File: leecode/82-删除排序链表中的重复元素Ⅱ.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def deleteDuplicates(self, head):
        if not head or not head.next:
            return head
        p = ListNode(0)
        p.next = head
        clone2 = p
        clone = head
        while clone and clone.next:
            if clone.val != clone.next.val:
                clone = clone.next
                clone2 = clone2.next
            else:
                while clone.next and clone.val == clone.next.val:
                    clone = clone.next
                clone2.next = clone.next
                clone = clone.next
        return p.next

File: leecode/1-500/1-100/82-删除排序链表中的重复元素Ⅱ.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* q = new ListNode(-101);
        q -> next = head;
        ListNode* p = q;
        while (q) {
            bool flag = false;
            ListNode* clone = q ->next;
            while(clone && clone->next && clone->val == clone->next->val) {
                clone = clone->next;
                flag = true;
            }
            if (flag)
                q->next = clone->next;
            else
                q = q -> next;
        }
        return p -> next;
    }
};